taskcat/s3_stage.py

```diff
@@ -0,0 +1,149 @@
+
+class S3APIResponse:
+    def __init__(self, x):
+        self._http_code = x['ResponseMetadata']['HTTPStatusCode']
+
+    @property
+    def ok(self):
+        if self._http_code == 200:
+            return True
+        return False
+
+class S3BucketCreatorException(Exception):
+    pass
+
+class S3BucketCreator:
+    def __init__(self, config: Config):
+        self.name = ""
+        self.public = False
+        self.tags = []
+        self.region = 'us-east-1'
+        self.sigv4 = True
+        self._config = config
+        self._c = None
+
+    @property
+    def acl(self):
+        return self._acl
+
+    @property
+    def policy(self):
+        return self._policy
+
+    def _create_in_region(self, region):
+        if region == 'us-east-1':
+            response = self._c.create_bucket(
+                ACL=self.acl,
+                Bucket=self.name
+            )
+        else:
+            response = self._c.create_bucket(
+                ACL=self.acl,
+                Bucket=self.name,
+                CreateBucketConfiguration={
+                    'LocationConstraint': region
+                }
+            )
+
+        return S3APIResponse(response)
+
+    def _return_sigv4_policy(self, bucket):
+        policy = F"""{
+            "Version": "2012-10-17",
+            "Statement": [
+                {
+                    "Sid": "Test",
+                    "Effect": "Deny",
+                    "Principal": "*",
+                    "Action": "s3:*",
+                    "Resource": "arn:aws:s3:::{bucket}/*",
+                    "Condition": {
+                         "StringEquals": {
+                               "s3:signatureversion": "AWS"
+                         }
+                    }
+                }
+            ]
+        }"""
+
+    def create(self):
+        # Verify bucket name length
+        if len(self.name) > self.config.s3_bucket.max_name_len:
+            raise S3BucketCreatorException(f"The bucket you provided [{config.s3_bucket.name}] is greater than {config.s3_bucket.max_name_len} characters.")
+
+        s3_client = config.client_factory.get('s3', region=config.default_region, s3v4=self.sigv4)
+
+        if not config.s3_bucket.name:
+
+            # Verify bucket exists.
+            try:
+                _ = s3_client.list_objects(Bucket=config.s3_bucket.name)
```

Comment:
adding a return here will allow you to get rid of the else on line 86, so that the main logic of this method is not nested. Combined with the previous comment this method could look like:

```python
   def create(self):
        if len(self.name) > self.config.s3_bucket.max_name_len:
            raise S3BucketCreatorException(f"The bucket you provided [{config.s3_bucket.name}] is greater than {config.s3_bucket.max_name_len} characters.")
        s3_client = config.client_factory.get('s3', region=config.default_region, s3v4=self.sigv4)

        if not config.s3_bucket.name:
             return self._assert_bucket_exists()

        auto_bucket_name = f"taskcat-{self.stack_prefix}-{self.name}-{self.uuid}".lower()
        log.info(f"Creating bucket {auto_bucket_name} in {self.region}")
        config.s3_bucket.auto = True

        self._create_bucket(args)
```